src/main/java/org/folio/service/financedata/FinanceDataService.java

```diff
@@ -45,18 +53,93 @@ public FinanceDataService(FundService fundService, BudgetService budgetService,
 
   public Future<Void> update(FyFinanceDataCollection entity, RequestContext requestContext) {
     if (CollectionUtils.isEmpty(entity.getFyFinanceData())) {
-      return Future.succeededFuture();
+      return succeededFuture();
     }
     var dbClient = requestContext.toDBClient();
+    Map<String, String> okapiHeaders = requestContext.getHeaders();
     return dbClient
-      .withTrans(conn -> {
-        var updateFundFuture = processFundUpdate(entity, conn);
-        var updateBudgetFuture = processBudgetUpdate(entity, conn);
-        return GenericCompositeFuture.all(List.of(updateFundFuture, updateBudgetFuture))
-          .compose(v -> processAllocationTransaction(entity, conn, requestContext.getHeaders()));
-      })
+      .withTrans(conn -> fiscalYearService.getFiscalYearById(getFiscalYearId(entity), conn)
+        .compose(fiscalYear -> createBudgets(entity, fiscalYear, conn, okapiHeaders)
+          .compose(v -> updateFundAndBudget(entity, conn))
+          .compose(v -> processAllocationTransaction(entity, fiscalYear, conn, okapiHeaders))))
       .onSuccess(v -> logger.info("Successfully updated finance data"))
-      .onFailure(e -> logger.error("Failed to update finance data", e))
+      .onFailure(e -> logger.error("Failed to update finance data", e));
+  }
+
+  private Future<Void> createBudgets(FyFinanceDataCollection entity, FiscalYear fiscalYear, DBConn conn,
+      Map<String, String> okapiHeaders) {
+    if (fiscalYearIsPast(fiscalYear)) {
+      return succeededFuture();
+    }
+    List<FyFinanceData> dataList = entity.getFyFinanceData().stream()
+      .filter(data -> data.getBudgetId() == null && (data.getBudgetStatus() != null ||
+        requireNonNullElse(data.getBudgetAllocationChange(), 0.0) > 0))
+      .toList();
+    if (dataList.isEmpty()) {
+      return succeededFuture();
+    }
+    List<Budget> budgets = dataList.stream()
+      .map(data -> {
+        String budgetId = UUID.randomUUID().toString();
+        Budget.BudgetStatus budgetStatus = newBudgetStatus(fiscalYear);
+        data.setBudgetId(budgetId);
+        if (data.getBudgetStatus() == null) {
+          data.setBudgetStatus(FyFinanceData.BudgetStatus.fromValue(budgetStatus.value()));
+        }
+        Budget budget = new Budget()
+          .withId(budgetId)
+          .withName(requireNonNullElse(data.getBudgetName(), data.getFundCode() + '-' + data.getFiscalYearCode()))
+          .withBudgetStatus(budgetStatus)
+          .withAllocated(0.)
+          .withFundId(data.getFundId())
+          .withFiscalYearId(data.getFiscalYearId());
+        populateBudgetMetadata(budget, okapiHeaders);
+        return budget;
+      })
+      .toList();
+    if (budgets.isEmpty()) {
+      return succeededFuture();
+    }
```

Comment:
If am not mistaken we can remove this check because we are not doing filter in previous operation

Reply:
OK